File: scripts/core/storage/runtime_state.py
import logging
from typing import Optional

from core.storage.connection import _get_conn, closing

def get_runtime_hook_value(session_id: str, turn_idx: int, key: str) -> Optional[str]:
    try:
        with closing(_get_conn()) as conn:
            with conn:
                row = conn.execute(
                    "SELECT value FROM runtime_hook_state WHERE session_id = ? AND turn_idx = ? AND key = ?",
                    (session_id, turn_idx, key)
                ).fetchone()
                return row[0] if row else None
    except Exception as e:
        logging.error(f"Error in get_runtime_hook_value: {e}")
        return None

def set_runtime_hook_value(session_id: str, turn_idx: int, key: str, value: str) -> None:
    try:
        with closing(_get_conn()) as conn:
            with conn:
                conn.execute("BEGIN EXCLUSIVE")
                conn.execute(
                    "INSERT INTO runtime_hook_state (session_id, turn_idx, key, value) VALUES (?, ?, ?, ?) "
                    "ON CONFLICT(session_id, turn_idx, key) DO UPDATE SET value = excluded.value",
                    (session_id, turn_idx, key, value)
                )
    except Exception as e:
        logging.error(f"Error in set_runtime_hook_value: {e}")

def delete_runtime_hook_value(session_id: str, turn_idx: int, key: str) -> None:
    try:
        with closing(_get_conn()) as conn:
            with conn:
                conn.execute("BEGIN EXCLUSIVE")
                conn.execute(
                    "DELETE FROM runtime_hook_state WHERE session_id = ? AND turn_idx = ? AND key = ?",
                    (session_id, turn_idx, key)
                )
    except Exception as e:
        logging.error(f"Error in delete_runtime_hook_value: {e}")

def trim_runtime_hook_states(session_id: str, current_turn_idx: int) -> None:
    try:
        with closing(_get_conn()) as conn:
            with conn:
                conn.execute("BEGIN EXCLUSIVE")
                conn.execute(
                    "DELETE FROM runtime_hook_state WHERE session_id = ? AND turn_idx >= ?",
                    (session_id, current_turn_idx)
                )
    except Exception as e:
        logging.error(f"Error in trim_runtime_hook_states: {e}")

def get_hook_state(session_id: str, turn_idx: int, key: str) -> Optional[str]:
    return get_runtime_hook_value(session_id, turn_idx, key)

def set_hook_state(session_id: str, turn_idx: int, key: str, value: str) -> None:
    set_runtime_hook_value(session_id, turn_idx, key, value)

def delete_hook_state(session_id: str, turn_idx: int, key: str) -> None:
    delete_runtime_hook_value(session_id, turn_idx, key)

def trim_hook_states(session_id: str, current_turn_idx: int) -> None:
    trim_runtime_hook_states(session_id, current_turn_idx)

File: packages/core/src/storage/runtime-state.ts
import Database from "better-sqlite3";
import { getConn } from "./connection";

export function getRuntimeHookValue(
  sessionId: string,
  turnIdx: number,
  key: string,
  conn?: Database.Database,
): string | null {
  const db = conn ?? getConn();
  const ownConn = !conn;
  try {
    const row = db
      .prepare(
        "SELECT value FROM runtime_hook_state WHERE session_id = ? AND turn_idx = ? AND key = ?"
      )
      .get(sessionId, turnIdx, key) as { value: string } | undefined;
    return row ? row.value : null;
  } catch (e) {
    console.warn(`getRuntimeHookValue: ${e}`);
    return null;
  } finally {
    if (ownConn) db.close();
  }
}

export function setRuntimeHookValue(
  sessionId: string,
  turnIdx: number,
  key: string,
  value: string,
  conn?: Database.Database,
): void {
  const db = conn ?? getConn();
  const ownConn = !conn;
  try {
    db.prepare("BEGIN EXCLUSIVE").run();
    db
      .prepare(
        "INSERT INTO runtime_hook_state (session_id, turn_idx, key, value) VALUES (?, ?, ?, ?) " +
          "ON CONFLICT(session_id, turn_idx, key) DO UPDATE SET value = excluded.value"
      )
      .run(sessionId, turnIdx, key, value);
    db.prepare("COMMIT").run();
  } catch (e) {
    console.warn(`setRuntimeHookValue: ${e}`);
  } finally {
    if (ownConn) db.close();
  }
}

export function deleteRuntimeHookValue(
  sessionId: string,
  turnIdx: number,
  key: string,
  conn?: Database.Database,
): void {
  const db = conn ?? getConn();
  const ownConn = !conn;
  try {
    db.prepare("BEGIN EXCLUSIVE").run();
    db
      .prepare(
        "DELETE FROM runtime_hook_state WHERE session_id = ? AND turn_idx = ? AND key = ?"
      )
      .run(sessionId, turnIdx, key);
    db.prepare("COMMIT").run();
  } catch (e) {
    console.warn(`deleteRuntimeHookValue: ${e}`);
  } finally {
    if (ownConn) db.close();
  }
}

export function trimRuntimeHookStates(
  sessionId: string,
  currentTurnIdx: number,
  conn?: Database.Database,
): void {
  const db = conn ?? getConn();
  const ownConn = !conn;
  try {
    db.prepare("BEGIN EXCLUSIVE").run();
    db
      .prepare(
        "DELETE FROM runtime_hook_state WHERE session_id = ? AND turn_idx >= ?"
      )
      .run(sessionId, currentTurnIdx);
    db.prepare("COMMIT").run();
  } catch (e) {
    console.warn(`trimRuntimeHookStates: ${e}`);
  } finally {
    if (ownConn) db.close();
  }
}

export function getHookState(
  sessionId: string,
  turnIdx: number,
  key: string,
  conn?: Database.Database,
): string | null {
  return getRuntimeHookValue(sessionId, turnIdx, key, conn);
}

export function setHookState(
  sessionId: string,
  turnIdx: number,
  key: string,
  value: string,
  conn?: Database.Database,
): void {
  setRuntimeHookValue(sessionId, turnIdx, key, value, conn);
}

export function deleteHookState(
  sessionId: string,
  turnIdx: number,
  key: string,
  conn?: Database.Database,
): void {
  deleteRuntimeHookValue(sessionId, turnIdx, key, conn);
}

export function trimHookStates(
  sessionId: string,
  currentTurnIdx: number,
  conn?: Database.Database,
): void {
  trimRuntimeHookStates(sessionId, currentTurnIdx, conn);
}
